File: android/app/src/main/java/com/workapp/hafaka/ui/Editors.kt
package com.workapp.hafaka.ui

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.DialogProperties
import com.workapp.hafaka.data.Store
import com.workapp.hafaka.model.*

/** A dialog that fills most of the screen — the Android answer to a sheet. */
@Composable
private fun EditorDialog(
    title: String,
    onDismiss: () -> Unit,
    onSave: (() -> Unit)? = null,
    saveEnabled: Boolean = true,
    content: @Composable ColumnScope.() -> Unit,
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(usePlatformDefaultWidth = false),
        modifier = Modifier.fillMaxWidth(0.96f).fillMaxHeight(0.9f),
        title = { Text(title, style = MaterialTheme.typography.titleLarge) },
        text = {
            Column(
                Modifier.fillMaxWidth().verticalScroll(rememberScrollState()),
                verticalArrangement = Arrangement.spacedBy(10.dp),
                content = content,
            )
        },
        confirmButton = {
            onSave?.let { TextButton(onClick = it, enabled = saveEnabled) { Text("שמירה") } }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text("ביטול") } },
    )
}

@Composable
private fun Field(
    label: String,
    value: String,
    onValueChange: (String) -> Unit,
    keyboard: KeyboardType = KeyboardType.Text,
    singleLine: Boolean = true,
    supporting: String? = null,
) {
    OutlinedTextField(
        value = value,
        onValueChange = onValueChange,
        label = { Text(label) },
        singleLine = singleLine,
        keyboardOptions = KeyboardOptions(keyboardType = keyboard),
        supportingText = supporting?.let { { Text(it) } },
        modifier = Modifier.fillMaxWidth(),
    )
}

// ---------------------------------------------------------------- person

@Composable
fun PersonEditor(store: Store, person: Person?, onDismiss: () -> Unit, onSaved: (Person) -> Unit = {}) {
    val isNew = person == null
    var draft by remember { mutableStateOf(person ?: Person()) }
    var confirmDelete by remember { mutableStateOf(false) }

    EditorDialog(
        title = if (isNew) "איש קשר חדש" else "עריכת איש קשר",
        onDismiss = onDismiss,
        saveEnabled = draft.name.isNotBlank(),
        onSave = { store.upsertPerson(draft); onSaved(draft); onDismiss() },
    ) {
        Field("שם מלא", draft.name, { draft = draft.copy(name = it) })
        Field("טלפון", draft.phone, { draft = draft.copy(phone = it) }, KeyboardType.Phone)
        DropdownField("מחלקה", draft.dept.he, Dept.entries.map { it.he }) { i ->
            draft = draft.copy(dept = Dept.entries[i])
        }
        DropdownField(
            "תפקיד קבוע",
            Sheets.slot(draft.defaultSlot)?.short ?: "ללא",
            listOf("ללא") + Sheets.all.map { it.short },
        ) { i -> draft = draft.copy(defaultSlot = if (i == 0) "" else Sheets.all[i - 1].slot) }
        Field("אימייל", draft.email, { draft = draft.copy(email = it) }, KeyboardType.Email)
        Field("יוצא מ־", draft.homeBase, { draft = draft.copy(homeBase = it) },
              supporting = "עיר או אזור — עוזר לתכנן איסופים")
        Field("הערות", draft.notes, { draft = draft.copy(notes = it) }, singleLine = false)

        if (!isNew) {
            TextButton(onClick = { confirmDelete = true }) {
                Text("מחיקת איש קשר", color = MaterialTheme.colorScheme.error)
            }
        }
    }

    if (confirmDelete) {
        AlertDialog(
            onDismissRequest = { confirmDelete = false },
            title = { Text("מחיקת איש קשר") },
            text = { Text("למחוק את ${draft.name}? הוא יוסר מכל ימי הצילום.") },
            confirmButton = {
                TextButton(onClick = { store.deletePerson(draft.id); confirmDelete = false; onDismiss() }) {
                    Text("מחק", color = MaterialTheme.colorScheme.error)
                }
            },
            dismissButton = { TextButton(onClick = { confirmDelete = false }) { Text("ביטול") } },
        )
    }
}

// ---------------------------------------------------------------- location

@Composable
fun LocationEditor(store: Store, location: Location?, onDismiss: () -> Unit, onSaved: (Location) -> Unit = {}) {
    val isNew = location == null
    var draft by remember { mutableStateOf(location ?: Location()) }
    var confirmDelete by remember { mutableStateOf(false) }

    EditorDialog(
        title = if (isNew) "מיקום חדש" else "עריכת מיקום",
        onDismiss = onDismiss,
        saveEnabled = draft.name.isNotBlank(),
        onSave = { store.upsertLocation(draft); onSaved(draft); onDismiss() },
    ) {
        Field("לוקיישן / סט", draft.name, { draft = draft.copy(name = it) })
        Field("כתובת", draft.address, { draft = draft.copy(address = it) })
        Field("חניה", draft.parking, { draft = draft.copy(parking = it) })
        // Coordinates stay text: an empty field must mean "not set", which a
        // numeric field would collapse to zero.
        Field("קו רוחב", draft.lat?.toString() ?: "",
              { draft = draft.copy(lat = it.replace(',', '.').toDoubleOrNull()) }, KeyboardType.Decimal,
              supporting = "לא חובה — בלי קואורדינטות הניווט מחפש לפי הכתובת")
        Field("קו אורך", draft.lng?.toString() ?: "",
              { draft = draft.copy(lng = it.replace(',', '.').toDoubleOrNull()) }, KeyboardType.Decimal)
        Field("הערות", draft.notes, { draft = draft.copy(notes = it) }, singleLine = false)

        if (!isNew) {
            TextButton(onClick = { confirmDelete = true }) {
                Text("מחיקת מיקום", color = MaterialTheme.colorScheme.error)
            }
        }
    }

    if (confirmDelete) {
        AlertDialog(
            onDismissRequest = { confirmDelete = false },
            title = { Text("מחיקת מיקום") },
            text = { Text("למחוק את המיקום \"${draft.name}\"?") },
            confirmButton = {
                TextButton(onClick = { store.deleteLocation(draft.id); confirmDelete = false; onDismiss() }) {
                    Text("מחק", color = MaterialTheme.colorScheme.error)
                }
            },
            dismissButton = { TextButton(onClick = { confirmDelete = false }) { Text("ביטול") } },
        )
    }
}

// ---------------------------------------------------------------- day

@Composable
fun DayEditor(store: Store, day: ShootDay?, onDismiss: () -> Unit, onSaved: (ShootDay) -> Unit = {}) {
    val isNew = day == null
    var draft by remember { mutableStateOf(day ?: ShootDay(date = HebDate.todayIso())) }
    var confirmDelete by remember { mutableStateOf(false) }
    val locations = store.locations

    EditorDialog(
        title = if (isNew) "יום צילום חדש" else "הגדרות היום",
        onDismiss = onDismiss,
        onSave = { store.upsertDay(draft); onSaved(draft); onDismiss() },
    ) {
        Field("תאריך (yyyy-MM-dd)", draft.date, { draft = draft.copy(date = it) },
              supporting = HebDate.long(draft.date))
        Field("כותרת", draft.title, { draft = draft.copy(title = it) },
              supporting = "למשל: יום 4 — סצנות 12-18")
        DropdownField(
            "מיקום",
            store.location(draft.locationId)?.name ?: "לא נבחר",
            listOf("לא נבחר") + locations.map { it.name },
        ) { i -> draft = draft.copy(locationId = if (i == 0) "" else locations[i - 1].id) }

        Field("קריאה כללית", draft.generalCall, { draft = draft.copy(generalCall = it) }, KeyboardType.Number)
        Field("תחילת צילום", draft.shootingCall, { draft = draft.copy(shootingCall = it) }, KeyboardType.Number)
        Field("סיום משוער", draft.wrap, { draft = draft.copy(wrap = it) }, KeyboardType.Number)
        Field("הערות הפקה", draft.notes, { draft = draft.copy(notes = it) }, singleLine = false)

        if (!isNew) {
            TextButton(onClick = { confirmDelete = true }) {
                Text("מחיקת יום", color = MaterialTheme.colorScheme.error)
            }
        }
    }

    if (confirmDelete) {
        AlertDialog(
            onDismissRequest = { confirmDelete = false },
            title = { Text("מחיקת יום צילום") },
            text = { Text("למחוק את יום הצילום הזה על כל הנתונים שבו?") },
            confirmButton = {
                TextButton(onClick = { store.deleteDay(draft.id); confirmDelete = false; onDismiss() }) {
                    Text("מחק", color = MaterialTheme.colorScheme.error)
                }
            },
            dismissButton = { TextButton(onClick = { confirmDelete = false }) { Text("ביטול") } },
        )
    }
}

// ---------------------------------------------------------------- call

@Composable
fun CallEditor(store: Store, day: ShootDay, person: Person, onDismiss: () -> Unit) {
    var draft by remember { mutableStateOf(day.calls[person.id] ?: CallInfo()) }
    val locations = store.locations

    EditorDialog(
        title = person.name,
        onDismiss = onDismiss,
        onSave = { store.setCall(day.id, person.id, draft); onDismiss() },
    ) {
        Field("שעת קריאה", draft.time, { draft = draft.copy(time = it) }, KeyboardType.Number,
              supporting = "ריק = הקריאה הכללית (${day.generalCall})")
        DropdownField(
            "מיקום",
            store.location(draft.locationId)?.name ?: "כמו היום",
            listOf("כמו היום") + locations.map { it.name },
        ) { i -> draft = draft.copy(locationId = if (i == 0) "" else locations[i - 1].id) }

        Text("סטטוס", style = MaterialTheme.typography.labelLarge)
        SingleChoiceSegmentedButtonRow(Modifier.fillMaxWidth()) {
            CallStatus.entries.forEachIndexed { i, s ->
                SegmentedButton(
                    selected = draft.status == s,
                    onClick = { draft = draft.copy(status = s) },
                    shape = SegmentedButtonDefaults.itemShape(i, CallStatus.entries.size),
                ) { Text(s.he, style = MaterialTheme.typography.labelMedium) }
            }
        }
        Field("הערה", draft.note, { draft = draft.copy(note = it) }, singleLine = false,
              supporting = "איסוף, ציוד מיוחד…")
    }
}

// ---------------------------------------------------------------- vehicle

@Composable
fun VehicleEditor(store: Store, day: ShootDay, vslot: String, onDismiss: () -> Unit) {
    var draft by remember { mutableStateOf(day.vehicles[vslot] ?: VehicleAssignment()) }
    var picking by remember { mutableStateOf(false) }
    val label = Sheets.vehicles.firstOrNull { it.first == vslot }?.second ?: "רכב"

    EditorDialog(
        title = label,
        onDismiss = onDismiss,
        onSave = {
            store.editDay(day.id) { d -> d.copy(vehicles = d.vehicles + (vslot to draft)) }
            onDismiss()
        },
    ) {
        val driver = store.person(draft.driverId)
        ListItem(
            headlineContent = { Text(driver?.name ?: "שיוך נהג") },
            supportingContent = {
                val phone = driver?.phone.orEmpty()
                if (phone.isNotBlank()) LtrText(Phone.pretty(phone))
            },
            leadingContent = {
                if (driver != null) Avatar(driver.name, 36, driver.dept.hex)
                else Avatar("+", 36, "#8E8E93")
            },
            modifier = Modifier.clickable { picking = true },
        )
        Field("מספר רכב", draft.plate, { draft = draft.copy(plate = it) })
        Field("הערה", draft.note, { draft = draft.copy(note = it) }, singleLine = false,
              supporting = "מה נוסע ברכב הזה")
    }

    if (picking) {
        PersonPicker(store, title = "נהג — $label", slot = null, selectedId = draft.driverId,
                     onDismiss = { picking = false }) { id -> draft = draft.copy(driverId = id) }
    }
}

// ---------------------------------------------------------------- picker

@Composable
fun PersonPicker(
    store: Store,
    title: String,
    slot: String?,
    selectedId: String? = null,
    allowNone: Boolean = true,
    onDismiss: () -> Unit,
    onPick: (String) -> Unit,
) {
    var query by remember { mutableStateOf("") }
    var creating by remember { mutableStateOf(false) }
    val dept = slot?.let { Sheets.slot(it)?.dept }

    // People whose default role matches this slot float to the top, then the
    // rest of the department, then everyone else.
    val ordered = remember(query, store.people, slot) {
        store.people
            .filter { query.isBlank() || "${it.name} ${it.phone} ${it.dept.he}".contains(query, true) }
            .sortedWith(compareBy(
                { p -> if (slot != null && p.defaultSlot == slot) 0 else if (dept != null && p.dept == dept) 1 else 2 },
                { it.name },
            ))
    }

    AlertDialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(usePlatformDefaultWidth = false),
        modifier = Modifier.fillMaxWidth(0.96f).fillMaxHeight(0.85f),
        title = { Text(title, style = MaterialTheme.typography.titleLarge) },
        text = {
            Column(Modifier.fillMaxSize()) {
                OutlinedTextField(
                    value = query, onValueChange = { query = it },
                    label = { Text("חיפוש") }, singleLine = true,
                    modifier = Modifier.fillMaxWidth(),
                )
                LazyColumn(Modifier.weight(1f)) {
                    if (allowNone) {
                        item {
                            ListItem(
                                headlineContent = { Text("להשאיר ריק") },
                                modifier = Modifier.clickable { onPick(""); onDismiss() },
                            )
                        }
                    }
                    items(ordered, key = { it.id }) { p ->
                        ListItem(
                            headlineContent = { Text(p.name) },
                            supportingContent = {
                                Text(listOfNotNull(p.dept.he, Sheets.slot(p.defaultSlot)?.short)
                                    .joinToString(" · "))
                            },
                            leadingContent = { Avatar(p.name, 38, p.dept.hex) },
                            trailingContent = {
                                when {
                                    slot != null && p.defaultSlot == slot -> Pill("ברירת מחדל")
                                    selectedId == p.id -> Text("✓", color = Tint)
                                    else -> Unit
                                }
                            },
                            modifier = Modifier.clickable { onPick(p.id); onDismiss() },
                        )
                    }
                    item {
                        TextButton(onClick = { creating = true }, modifier = Modifier.padding(8.dp)) {
                            Text("+ איש קשר חדש")
                        }
                    }
                }
            }
        },
        confirmButton = { TextButton(onClick = onDismiss) { Text("סגור") } },
    )

    if (creating) {
        PersonEditor(
            store,
            person = Person(dept = dept ?: Dept.PRODUCTION, defaultSlot = slot.orEmpty()),
            onDismiss = { creating = false },
            onSaved = { p -> onPick(p.id); onDismiss() },
        )
    }
}

// ---------------------------------------------------------------- dropdown

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun DropdownField(label: String, selected: String, options: List<String>, onSelect: (Int) -> Unit) {
    var expanded by remember { mutableStateOf(false) }
    ExposedDropdownMenuBox(expanded = expanded, onExpandedChange = { expanded = it }) {
        OutlinedTextField(
            value = selected,
            onValueChange = {},
            readOnly = true,
            label = { Text(label) },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded) },
            modifier = Modifier.fillMaxWidth().menuAnchor(MenuAnchorType.PrimaryNotEditable),
        )
        ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            options.forEachIndexed { i, option ->
                DropdownMenuItem(
                    text = { Text(option) },
                    onClick = { onSelect(i); expanded = false },
                )
            }
        }
    }
}

File: ios/WorkApp/Views/Editors.swift
import SwiftUI
import CoreLocation
import Combine

// MARK: - Person

struct PersonEditor: View {
    @EnvironmentObject var store: Store
    @Environment(\.dismiss) private var dismiss

    @State var draft: Person
    var isNew: Bool
    var onSave: ((Person) -> Void)?

    @State private var confirmDelete = false

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    TextField("שם מלא", text: $draft.name)
                    TextField("050-000-0000", text: $draft.phone)
                        .keyboardType(.phonePad)
                    Picker("מחלקה", selection: $draft.dept) {
                        ForEach(Dept.allCases) { Text($0.he).tag($0) }
                    }
                    Picker("תפקיד קבוע", selection: $draft.defaultSlot) {
                        Text("ללא").tag("")
                        ForEach(Sheets.all) { Text($0.short).tag($0.slot) }
                    }
                }

                Section {
                    TextField("אימייל (לא חובה)", text: $draft.email)
                        .keyboardType(.emailAddress)
                        .textInputAutocapitalization(.never)
                    TextField("יוצא מ־ (עיר / אזור)", text: $draft.homeBase)
                } footer: {
                    Text("״יוצא מ־״ עוזר לתכנן איסופים ולהעריך זמני נסיעה.")
                }

                Section("הערות") {
                    TextField("ציוד, העדפות, מגבלות…", text: $draft.notes, axis: .vertical)
                        .lineLimit(3...6)
                }

                if !isNew {
                    Section {
                        Button("מחיקת איש קשר", role: .destructive) { confirmDelete = true }
                    }
                }
            }
            .navigationTitle(isNew ? "איש קשר חדש" : "עריכת איש קשר")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("ביטול") { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button("שמירה") {
                        store.upsert(draft)
                        onSave?(draft)
                        dismiss()
                    }
                    .disabled(draft.name.trimmingCharacters(in: .whitespaces).isEmpty)
                }
            }
            .confirmationDialog("למחוק את \(draft.name)? הוא יוסר מכל ימי הצילום.",
                                isPresented: $confirmDelete, titleVisibility: .visible) {
                Button("מחק", role: .destructive) {
                    store.deletePerson(draft.id)
                    dismiss()
                }
            }
        }
    }
}

// MARK: - Location

struct LocationEditor: View {
    @EnvironmentObject var store: Store
    @Environment(\.dismiss) private var dismiss

    @State var draft: Location
    var isNew: Bool
    var onSave: ((Location) -> Void)?

    @StateObject private var locator = OneShotLocator()
    @State private var confirmDelete = false

    /// Coordinates are optional, and an empty field has to stay empty rather
    /// than collapsing to 0 — which rules out the numeric TextField overloads.
    private func coordinateField(_ label: String, _ placeholder: String,
                                 _ value: Binding<Double?>) -> some View {
        let text = Binding<String>(
            get: { value.wrappedValue.map { String($0) } ?? "" },
            set: { value.wrappedValue = Double($0.replacingOccurrences(of: ",", with: ".")) })
        return HStack {
            Text(label)
            Spacer()
            TextField(placeholder, text: text)
                .keyboardType(.decimalPad)
                .multilineTextAlignment(.trailing)
                .frame(width: 130)
                .environment(\.layoutDirection, .leftToRight)
        }
    }

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    TextField("לוקיישן / סט", text: $draft.name)
                    TextField("רחוב, עיר", text: $draft.address)
                    TextField("איפה חונים", text: $draft.parking)
                }

                Section {
                    coordinateField("קו רוחב", "32.0853", $draft.lat)
                    coordinateField("קו אורך", "34.7818", $draft.lng)
                    Button {
                        locator.request()
                    } label: {
                        Label(locator.isWorking ? "מאתר מיקום…" : "השתמש במיקום הנוכחי שלי",
                              systemImage: "location.fill")
                    }
                    .disabled(locator.isWorking)
                } footer: {
                    Text("קואורדינטות הן לא חובה. בלעדיהן הניווט מחפש לפי הכתובת; איתן הניווט מדויק גם בשטח פתוח בלי כתובת.")
                }

                Section("הערות") {
                    TextField("גישה, מפתחות, איש קשר בשטח…", text: $draft.notes, axis: .vertical)
                        .lineLimit(3...6)
                }

                if !isNew {
                    Section {
                        Button("מחיקת מיקום", role: .destructive) { confirmDelete = true }
                    }
                }
            }
            .navigationTitle(isNew ? "מיקום חדש" : "עריכת מיקום")
            .navigationBarTitleDisplayMode(.inline)
            .onReceive(locator.$coordinate.compactMap { $0 }) { coord in
                draft.lat = coord.latitude
                draft.lng = coord.longitude
            }
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("ביטול") { dismiss() } }
                ToolbarItem(placement: .confirmationAction) {
                    Button("שמירה") {
                        store.upsert(draft)
                        onSave?(draft)
                        dismiss()
                    }
                    .disabled(draft.name.trimmingCharacters(in: .whitespaces).isEmpty)
                }
            }
            .confirmationDialog("למחוק את המיקום \"\(draft.name)\"?",
                                isPresented: $confirmDelete, titleVisibility: .visible) {
                Button("מחק", role: .destructive) {
                    store.deleteLocation(draft.id)
                    dismiss()
                }
            }
        }
    }
}

// MARK: - Person picker (fills a sheet slot)

struct PersonPicker: View {
    @EnvironmentObject var store: Store
    @Environment(\.dismiss) private var dismiss

    var title: String
    var slot: String?
    var selectedId: String?
    var allowNone: Bool = true
    var onPick: (String) -> Void

    @State private var query = ""
    @State private var creating = false

    /// People whose default role matches this slot float to the top, then the
    /// rest of the department, then everyone else.
    private var ordered: [Person] {
        let dept = slot.flatMap { Sheets.slot($0)?.dept }
        let filtered = query.isEmpty ? store.people : store.people.filter {
            "\($0.name) \($0.phone) \($0.dept.he)".localizedCaseInsensitiveContains(query)
        }
        return filtered.sorted { a, b in
            func rank(_ p: Person) -> Int {
                if let slot, p.defaultSlot == slot { return 0 }
                if let dept, p.dept == dept { return 1 }
                return 2
            }
            return rank(a) == rank(b)
                ? a.name.localizedCompare(b.name) == .orderedAscending
                : rank(a) < rank(b)
        }
    }

    var body: some View {
        NavigationStack {
            List {
                if allowNone {
                    Button {
                        onPick("")
                        dismiss()
                    } label: {
                        HStack {
                            Text("להשאיר ריק").foregroundStyle(.primary)
                            Spacer()
                            if selectedId?.isEmpty ?? true {
                                Image(systemName: "checkmark").foregroundStyle(Color.accentColor)
                            }
                        }
                    }
                }

                ForEach(ordered) { person in
                    Button {
                        onPick(person.id)
                        dismiss()
                    } label: {
                        HStack(spacing: 12) {
                            AvatarView(name: person.name, size: 38, hex: person.dept.hex)
                            VStack(alignment: .leading, spacing: 1) {
                                Text(person.name).foregroundStyle(.primary)
                                Text([person.dept.he,
                                      Sheets.slot(person.defaultSlot)?.short]
                                        .compactMap { $0 }.joined(separator: " · "))
                                    .font(.footnote)
                                    .foregroundStyle(.secondary)
                            }
                            Spacer()
                            if person.defaultSlot == slot, slot != nil {
                                Pill(text: "ברירת מחדל")
                            }
                            if selectedId == person.id {
                                Image(systemName: "checkmark").foregroundStyle(Color.accentColor)
                            }
                        }
                    }
                }

                Section {
                    Button {
                        creating = true
                    } label: {
                        Label("איש קשר חדש", systemImage: "plus")
                    }
                }
            }
            .searchable(text: $query, prompt: "חיפוש")
            .navigationTitle(title)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("ביטול") { dismiss() } }
            }
            .sheet(isPresented: $creating) {
                PersonEditor(
                    draft: Person(dept: slot.flatMap { Sheets.slot($0)?.dept } ?? .production,
                                  defaultSlot: slot ?? ""),
                    isNew: true,
                    onSave: { person in
                        onPick(person.id)
                        dismiss()
                    })
            }
        }
    }
}

// MARK: - Per-person call time and status

struct CallEditor: View {
    @EnvironmentObject var store: Store
    @Environment(\.dismiss) private var dismiss

    let day: ShootDay
    let person: Person
    @State var draft: CallInfo

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    TextField("שעת קריאה (HH:MM)", text: $draft.time)
                        .keyboardType(.numbersAndPunctuation)
                    Picker("מיקום", selection: $draft.locationId) {
                        Text("כמו היום").tag("")
                        ForEach(store.locations) { Text($0.name).tag($0.id) }
                    }
                } footer: {
                    Text("קריאה כללית ליום זה: \(day.generalCall). השאירו ריק כדי להשתמש בה.")
                }

                Section("סטטוס") {
                    Picker("סטטוס", selection: $draft.status) {
                        ForEach(CallStatus.allCases) { Text($0.he).tag($0) }
                    }
                    .pickerStyle(.segmented)
                }

                Section("הערה") {
                    TextField("איסוף, ציוד מיוחד…", text: $draft.note, axis: .vertical)
                        .lineLimit(2...5)
                }

                Section {
                    Button {
                        Actions.call(person)
                    } label: { Label("חיוג", systemImage: "phone.fill") }
                    Button {
                        Actions.whatsapp(person, text: Actions.callMessage(
                            day: day, person: person, call: draft,
                            location: store.location(draft.locationId.isEmpty ? day.locationId : draft.locationId)))
                    } label: { Label("שליחת קריאה בוואטסאפ", systemImage: "message.fill") }
                }
            }
            .navigationTitle(person.name)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("ביטול") { dismiss() } }
                ToolbarItem(placement: .confirmationAction) {
                    Button("שמירה") {
                        store.setCall(day: day.id, person: person.id, draft)
                        dismiss()
                    }
                }
            }
        }
    }
}

// MARK: - Shoot day

struct DayEditor: View {
    @EnvironmentObject var store: Store
    @Environment(\.dismiss) private var dismiss

    @State var draft: ShootDay
    var isNew: Bool
    var onSave: ((ShootDay) -> Void)?

    @State private var date = Date()
    @State private var addingLocation = false
    @State private var confirmDelete = false

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    DatePicker("תאריך", selection: $date, displayedComponents: .date)
                    TextField("יום 4 — סצנות 12-18", text: $draft.title)
                    Picker("מיקום", selection: $draft.locationId) {
                        Text("לא נבחר").tag("")
                        ForEach(store.locations) { Text($0.name).tag($0.id) }
                    }
                    Button {
                        addingLocation = true
                    } label: { Label("מיקום חדש", systemImage: "plus") }
                }

                Section("שעות") {
                    TextField("קריאה כללית", text: $draft.generalCall)
                        .keyboardType(.numbersAndPunctuation)
                    TextField("תחילת צילום", text: $draft.shootingCall)
                        .keyboardType(.numbersAndPunctuation)
                    TextField("סיום משוער", text: $draft.wrap)
                        .keyboardType(.numbersAndPunctuation)
                }

                Section("הערות הפקה") {
                    TextField("מזג אוויר, ציוד מיוחד, שינויים…", text: $draft.notes, axis: .vertical)
                        .lineLimit(3...6)
                }

                if !isNew {
                    Section {
                        Button("מחיקת יום", role: .destructive) { confirmDelete = true }
                    }
                }
            }
            .navigationTitle(isNew ? "יום צילום חדש" : "הגדרות היום")
            .navigationBarTitleDisplayMode(.inline)
            .onAppear { date = HebDate.parse(draft.date) ?? Date() }
            .onChange(of: date) { _, new in draft.date = ShootDay.isoToday(new) }
            .sheet(isPresented: $addingLocation) {
                LocationEditor(draft: Location(), isNew: true) { loc in
                    draft.locationId = loc.id
                }
            }
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("ביטול") { dismiss() } }
                ToolbarItem(placement: .confirmationAction) {
                    Button("שמירה") {
                        store.upsert(draft)
                        onSave?(draft)
                        dismiss()
                    }
                }
            }
            .confirmationDialog("למחוק את יום הצילום הזה על כל הנתונים שבו?",
                                isPresented: $confirmDelete, titleVisibility: .visible) {
                Button("מחק", role: .destructive) {
                    store.deleteDay(draft.id)
                    dismiss()
                }
            }
        }
    }
}

// MARK: - Vehicle (sheet "רכבים")

struct VehicleEditor: View {
    @EnvironmentObject var store: Store
    @Environment(\.dismiss) private var dismiss

    let day: ShootDay
    let vslot: String
    @State var draft: VehicleAssignment

    @State private var pickingDriver = false

    private var label: String {
        Sheets.vehicles.first { $0.slot == vslot }?.he ?? "רכב"
    }

    var body: some View {
        NavigationStack {
            Form {
                Section("נהג") {
                    Button {
                        pickingDriver = true
                    } label: {
                        HStack(spacing: 12) {
                            if let p = store.person(draft.driverId) {
                                AvatarView(name: p.name, size: 36, hex: p.dept.hex)
                                VStack(alignment: .leading, spacing: 1) {
                                    Text(p.name).foregroundStyle(.primary)
                                    if !p.phone.isEmpty {
                                        LTRText(text: Actions.pretty(p.phone), font: .footnote)
                                            .foregroundStyle(.secondary)
                                    }
                                }
                            } else {
                                Circle().fill(Color(.tertiarySystemFill))
                                    .frame(width: 36, height: 36)
                                    .overlay(Image(systemName: "plus").foregroundStyle(.tertiary))
                                Text("שיוך נהג").foregroundStyle(.primary)
                            }
                            Spacer()
                            Image(systemName: "chevron.forward").foregroundStyle(.tertiary)
                        }
                    }
                }

                Section {
                    TextField("מספר רכב", text: $draft.plate)
                        .keyboardType(.numbersAndPunctuation)
                    TextField("מה נוסע ברכב הזה…", text: $draft.note, axis: .vertical)
                        .lineLimit(2...4)
                }
            }
            .navigationTitle(label)
            .navigationBarTitleDisplayMode(.inline)
            .sheet(isPresented: $pickingDriver) {
                PersonPicker(title: "נהג — \(label)", slot: nil, selectedId: draft.driverId) { id in
                    draft.driverId = id
                }
            }
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("ביטול") { dismiss() } }
                ToolbarItem(placement: .confirmationAction) {
                    Button("שמירה") {
                        store.editDay(day.id) { $0.vehicles[vslot] = draft }
                        dismiss()
                    }
                }
            }
        }
    }
}
